tools/integration-test/src/tests/client_upgrade.rs

```diff
@@ -0,0 +1,577 @@
+
+        let client_upgrade_height = Height::new(
+            foreign_clients.client_a_to_b.src_chain().id().version(),
+            upgrade_height,
+        )
+        .unwrap();
+
+        // Vote on the proposal so the chain will upgrade
+        vote_proposal(
+            driver.chain_id.as_str(),
+            &driver.command_path,
+            &driver.home_path,
+            &driver.rpc_listen_address(),
+        )?;
+
+        // The application height reports a height of 1 less than the height according to Tendermint
+        let target_reference_application_height = client_upgrade_height
+            .decrement()
+            .expect("Upgrade height cannot be 1");
+
+        let mut reference_application_latest_height = foreign_clients
+            .client_a_to_b
+            .src_chain()
+            .query_latest_height()
+            .ok()
+            .unwrap();
+
+        while reference_application_latest_height != target_reference_application_height {
+            std::thread::sleep(Duration::from_millis(500));
+
+            reference_application_latest_height = foreign_clients
+                .client_a_to_b
+                .src_chain()
+                .query_latest_height()
+                .ok()
+                .unwrap();
+        }
+
+        // Wait for the chain to upgrade
+        std::thread::sleep(WAIT_CHAIN_UPGRADE);
+
+        // Trigger the client upgrade
+        let outcome = foreign_clients.client_a_to_b.upgrade(client_upgrade_height);
+
+        assert!(outcome.is_ok(), "{:?}", outcome);
+
+        Ok(())
+    }
+}
+
+struct InvalidClientUpgradeTest;
+
+/// The genesis file doesn't need to be updated since the chain isn't upgraded
+/// for this test.
+impl TestOverrides for InvalidClientUpgradeTest {}
+
+impl BinaryChainTest for InvalidClientUpgradeTest {
+    fn run<
+        ChainA: ibc_test_framework::prelude::ChainHandle,
+        ChainB: ibc_test_framework::prelude::ChainHandle,
+    >(
+        &self,
+        _config: &ibc_test_framework::prelude::TestConfig,
+        _relayer: ibc_test_framework::prelude::RelayerDriver,
+        chains: ibc_test_framework::prelude::ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), ibc_test_framework::prelude::Error> {
+        let arbitrary_height = 5u64;
+        let foreign_clients = chains.foreign_clients;
+
+        // Wait for the chain to reach a given height
+        let client_upgrade_height = Height::new(
+            foreign_clients.client_a_to_b.src_chain().id().version(),
+            arbitrary_height,
+        )
+        .unwrap();
+
+        // Wait a bit before trying to upgrade the client
+        std::thread::sleep(Duration::from_secs(2));
+
+        // Trigger the client upgrade
+        let outcome = foreign_clients.client_a_to_b.upgrade(client_upgrade_height);
+
+        assert!(outcome.is_err(), "{:?}", outcome);
+
+        Ok(())
+    }
+}
+
+struct HeightTooHighClientUpgradeTest;
+
+impl TestOverrides for HeightTooHighClientUpgradeTest {
```

Comment:
If there are multiple identical test overrides, we can group them into a common override and implement `HasOverrides` instead. See `ExpirationTestOverrides` in `tools/integration-test/src/tests/client_expiration.rs` for example.

Reply:
This has been addressed in this commit https://github.com/informalsystems/hermes/pull/2795/commits/3e6b163b86f05455777d928039875cdcd28260ba